### PR TITLE
Group Journal entries

### PR DESCRIPTION
**Pull request by aumayr**
This PR implements #543: "Group journal entries by day, week, month"

Progress:

- [X] Add template filter
- [ ] Add tests
- [x] Modify JS-code for sorting the table

### CODE REVIEW

**Comment by yagebu** [comment]
Cool! Some preliminary notes:

- I think this could be enabled by default (and doesn't even need option to disable it IMHO).
- For the grouping interval, I'd just use `interval`.
- The grouping filter is highly inefficient and it adds ~1s for me - the entries are already sorted by group so there's not really any need for it. Just check for each entry if we switched group.

**Comment by aumayr** [comment]
I'm not quite happy with the current state:

- Sorting does not work (and would require big JS effort to implement).
- Hiding empty groups is done in JS, which leads to the rendering of empty groups until the whole journal table is rendered. This looks ugly.

@yagebu Do you have any ideas/suggestions on how to implement this in a better and more efficient way?

File: fava/template_filters.py [hunk 1]
```diff
@@ -173,20 +173,18 @@ def format_errormsg(message):
...
-        (x, [entry for entry in entries if group_fn(entry.date) == x])
-        for x in values]
+def date_group(entry):
+    """Return the date group an entry belongs to."""
+    kind = g.ledger.fava_options['interval']
+    d = entry.date
+    if kind == 'year':
+        return d.year
+    elif kind == 'quarter':
+        return '{}-Q{}'.format(d.year, (d.month-1)//3+1)
```
Comment (yagebu) [inline fava/template_filters.py]: there's no need to render this to a string, just return a tuple

Comment (yagebu) [reply fava/template_filters.py]: or maybe nevermind: for the comparison at least it's not needed, but to render out the headers it is. Might not be worth the time saved to have two functions for this

**Comment by yagebu** [comment]
- Don't hide empty groups. The fact that there aren't any entries shown in a group is a fact worth showing IMHO
- Sorting and grouping don't make sense together except for the initial order. So to sort, simply delete all group headers and then sort as before.

**Comment by aumayr** [comment]
> - Don't hide empty groups. The fact that there aren't any entries shown in a group is a fact worth showing IMHO

This is not really useful, as for every query/filter you set, all `open` statements will be in the Journal, and when hidden by default, empty groups for all `open` statements will be shown.

> Sorting and grouping don't make sense together except for the initial order. So to sort, simply delete all group headers and then sort as before.

Agreed, and implemented in 9e5e2af

**Comment by aumayr** [comment]
@yagebu What do you think of the current implementation (code-/design-wise)?

File: fava/template_filters.py [hunk 2]
```diff
@@ -171,3 +171,20 @@ def format_errormsg(message):
     new_message = re.sub(ACCOUNT_RE,
                          '<a href="{}">'.format(url) + r'\1' + '</a>', message)
     return new_message.replace('for \'', 'for ').replace('\': ', ': ')
+
+
+def date_group(entry):
+    """Return the date group an entry belongs to."""
+    kind = g.ledger.fava_options['interval']
```
Comment (yagebu) [inline fava/template_filters.py]: not the option, but the current value needs to be used

File: fava/template_filters.py [hunk 3]
```diff
@@ -171,3 +171,20 @@ def format_errormsg(message):
...
+    date = entry.date
+    if kind == 'year':
+        return date.year
+    elif kind == 'quarter':
+        return '{}-Q{}'.format(date.year, (date.month-1)//3+1)
+    elif kind == 'month':
+        return '{}-{:02}'.format(date.year, date.month)
```
Comment (yagebu) [inline fava/template_filters.py]: just return the tuple, same for the other cases

File: fava/templates/_journal_table.html [hunk 4]
```diff
@@ -92,100 +92,109 @@
...
+  {% set type = entry.__class__.__name__.lower() %}
+  {% set entry_hash = entry|hash_entry %}
+  {% set metadata = entry.meta|remove_keys(['__tolerances__', '__automatic__', 'filename', 'lineno']) %}
+  <li class="{{ type }} {{ entry.type or '' }} {{ entry.flag|flag_to_type if entry.flag else '' }}
+  {{- ' statement' if entry.tags and 'statement' in entry.tags else '' }}
+  {{- ' discovered' if entry.tags and 'discovered' in entry.tags else '' }}
+  {{- ' hidden' if not entry|show_journal_entry else '' }}
+  {{- ' fail' if entry.diff_amount else '' }}">
+    <p>
+      <span class="datecell" data-group-value="{{ ns.group }}" data-sort-value=-{{ loop.index }}><a href="#context-{{ entry_hash }}">{{ entry.date }}</a></span>
```
Comment (yagebu) [inline fava/templates/_journal_table.html]: unused data-group-value - remove

File: fava/template_filters.py (hunk 3 above)
Comment (aumayr) [reply fava/template_filters.py]: https://github.com/beancount/fava/pull/620#discussion_r159632840 ?

**Comment by yagebu** [comment]
I think overall that the implementation problems this creates make this somewhat not worth it. What do you think @aumayr? Are you planning to continue this (or probably redo since this has gotten somewhat unmergeable in parts)?

**Comment by aumayr** [comment]
Implementing (actually: redoing) this is still worth it I think, because at least once a month I wish for grouping by month. Will push this up on my todo list.

**Comment by yagebu** [comment]
Will close the PR for now, the issue's still open ;)

**Pull request closed without merging**